src/layers/basic.jl

```diff
@@ -428,3 +429,50 @@ function Base.show(io::IO, m::Parallel)
   join(io, m.layers, ", ")
   print(io, ")")
 end
+
+"""
+    Embedding(in => out; init=randn)
+
+A lookup table that stores embeddings of dimension `out` 
+for a vocabulary of size `in`. 
+
+This layers is often used to store word embeddings and retrieve them using indices. 
+The input to the layer can be either a vector of indexes
+or the corresponding [onehot encoding](@ref Flux.OneHotArray). 
+
+# Examples
+
+```jldoctest
+julia> m = Embedding(reshape(-6:45, 2, 26) .+ 0.01f0)
+Embedding(26 => 2)
+
+julia> m(5)  # embedding vector for 5th element
+2-element Vector{Float32}:
+ 2.01
+ 3.01
+
+julia> m([6, 15, 15])  # applied to a batch
+2×3 Matrix{Float32}:
+ 4.01  22.01  22.01
+ 5.01  23.01  23.01
+
+julia> ans == m(Flux.OneHotMatrix([6, 15, 15], 26))
+true
+```
+"""
+struct Embedding{W <: AbstractMatrix}
+  weight::W
+end
+
+@functor Embedding
+
+Embedding(dims::Pair{<:Integer, <:Integer}; init = randn32) = Embedding(init(last(dims), first(dims)))
+  
+(m::Embedding)(x::Union{OneHotVector, OneHotMatrix}) = m.weight * x # equivalent to m.weight[:,onecold(x)]
+(m::Embedding)(x::Integer) = m([x])
+(m::Embedding)(x::AbstractVector) = NNlib.gather(m.weight, x)
+(m::Embedding)(x::AbstractArray) = reshape(m(vec(x)), :, size(x)...)
```

Comment:
I would also restrict these to integers
```suggestion
(m::Embedding)(x::AbstractVector{<:Integer}) = NNlib.gather(m.weight, x)
(m::Embedding)(x::AbstractArray{<:Integer}) = reshape(m(vec(x)), :, size(x)...)
```

Reply:
Maybe only for `AbstractVector` so that `Nil`-arrays can go through the reshape?

Reply:
Oh I wasn't thinking about Nil, good point. Do you want Array{Nil} to behave like Array{Int} or like Array{Bool}? 

One other weird feature is that if you make a OneHotTensor, like a boolean 3-array, this will reshape it to a vector. Maybe that's too weird to worry about. Or maybe it should reshape `AbstractArray{Bool}` not with vec but with `m(reshape(size(x,1), :))`

Reply:
Yes, I think you are correct. Reshaping should be different when the first dimension encodes one-hot information. This definitely presents a problem for `Nil`, because the `Nil`-array is agnostic of whether the original input is one-hot or indices. In other words, the output size is input type dependent for this layer.

Reply:
In that case, is the only way for getting the correct output from `outputsize` just to ignore the first dimension of the one-hot input while passing in the `inputsizes` argument?